Stretchy HTML5 iframe height

kollivier opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

### Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

For some HTML5 content, there is enough height to render the content but the hardcoded size causes the content to be cut off and thus necessitates vertical scrollbars in the HTML5 content itself. This PR makes it so that the content can expand up to 70% of the page's virtual height, while retaining the old hardcoded height as the min-height to minimize issues with existing content.

### Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

ProFuturo content in the Science section is a good test of what this fixes, but testing with a variety of HTML5 content is ideal. We may want this to have another more extensive HTML5 QA pass before merging into develop.

### References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

…

----

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

kollivier committed:
Add CONTENTLOADED event so we can send the ideal height to use to render the iframe content to the HTML5 renderer.

kollivier committed:
Comment size fix.

kollivier committed:
Linter changes unrelated to my PR.

indirectlylit commented on kolibri/plugins/html5_viewer/assets/src/views/Html5AppRendererIndex.vue:
```diff
@@ -73,6 +73,15 @@
       this.pollProgress();
+      let self = this;
+      this.hashi.on(events.CONTENTLOADED, function(data) {
+        // The HTML5 renderer has min and max heights set, so exceedingly low or height height values
```
```suggestion
        // The HTML5 renderer has min and max heights set, so exceedingly low or high height values
```

indirectlylit commented on kolibri/plugins/html5_viewer/assets/src/views/Html5AppRendererIndex.vue:
```diff
@@ -114,7 +123,8 @@
     position: relative;
-    height: 500px;
+    min-height: 500px;
+    max-height: 70vh;
```
`vh` has [mediocre support](https://caniuse.com/#feat=viewport-units) on older iOS and Android devices.

We can do this, but if there is a way to avoid `vh` that would be preferable.

kollivier replied:
Any ideas on alternatives? Doing percentage would make it relative to parent element, rather than expanding the height based on available screen space, wouldn't it?

indirectlylit replied:
I'll take a look when I actually test it locally

indirectlylit commented on packages/hashi/src/iframeClient.js:
```diff
@@ -41,6 +41,15 @@ export default class SandboxEnvironment {
 
     patchCrossOrigin();
 
+    let self = this;
+    window.addEventListener('DOMContentLoaded', function() {
+      const data = {
+        offsetHeight: window.document.body.offsetHeight,
+      };
+      const event = events.CONTENTLOADED;
+      self.mediator.sendMessage({ nameSpace, event: event, data: data });
+    });
```
is this file packaged with Webpack, or is it unprocessed because it gets embedded in Hashi?

If the former, you can use an arrow function here:

```suggestion
    window.addEventListener('DOMContentLoaded', () => {
      const data = {
        offsetHeight: window.document.body.offsetHeight,
      };
      const event = events.CONTENTLOADED;
      this.mediator.sendMessage({ nameSpace, event: event, data: data });
    });
```

kollivier replied:
It's packaged, I'll make the change! Didn't realize this got around the self = this problem.

indirectlylit replied:
Yup - super convenient!

ref: https://hackernoon.com/javascript-es6-arrow-functions-and-lexical-this-f2a3e2a5e8c4

indirectlylit commented on kolibri/plugins/html5_viewer/assets/src/views/Html5AppRendererIndex.vue:
```diff
@@ -73,6 +73,15 @@
       this.$emit('startTracking');
       this.startTime = now();
       this.pollProgress();
+      let self = this;
+      this.hashi.on(events.CONTENTLOADED, function(data) {
```
`let self = this;` not necessary if you use an arrow function:

```suggestion
      this.hashi.on(events.CONTENTLOADED, (data) => {
```

indirectlylit commented on kolibri/plugins/html5_viewer/assets/src/views/Html5AppRendererIndex.vue:
```diff
@@ -73,6 +73,15 @@
+        // will not lead to problematic rendering scenarios. Consequently, this also means that it is only
+        // giving us the ability to stretch at most a couple hundred pixels height-wise.
+        if (data.offsetHeight && data.offsetHeight > 0) {
+          self.$refs.html5Renderer.$refs.fullscreen.style.height = data.offsetHeight + 'px';
```
rather than using nested `$refs` and modifying `.style.height`, it would be a bit more conventional to add a prop to `CoreFullscreen` and pass in a `height` if possible.

For example:

```html
<template>
  <div
    ref="fullscreen"
    :class="fullscreenClass"
    :style="{ height: `${height} px` }"
    allowfullscreen
  >
    <slot></slot>
  </div>
</template>

<script>
  // ...
    props: {
      height: { type: Number, required: false },
    }
   // ...
```

indirectlylit reviewed (commented):
will test soon - just a quick review of the code itself

indirectlylit commented:
I've been experimenting with this a bit on different content and seeing some odd results.

In `ProFuturo (English) / Level Up / Motivation` the height increases, but the width also does which causes horizontal scrolling:

![2020-03-19 18 21 24](https://user-images.githubusercontent.com/2367265/77128738-7d26c280-6a0e-11ea-9c1e-a4c976efa143.gif)

For other content such as PraDigi and HP Life, the iframe doesn't resize at all, presumably because the content is rendered after the `DOMContentLoaded` event.

My recommendation would be to something like:

* update the iFrame client to poll maybe once per second for content size changes
* send content size change events through Hashi when the size changes
* if necessary, incorporate both width and height information to avoid horizontal scrolling
* with regard to the `max-height: 70vh;` question, perhaps we don't need a max height at all? It might be better to _never_ scroll?

kollivier commented:
There are too many cases to handle fully dynamic sizing, I think. There are a number of different layout strategies a page may use:

1. Fixed height
2. Scale by aspect ratio
3. Fixed min-height but scales by aspect ratio
4. Small / med / large canvas sizes, chosen based on container height
5. Some completely custom sizing logic written in code

The problem is that the only way to know for sure which it is would be to check the code.

Polling and resizing every second will lead to recursion for cases 2 and sometimes 3 and 5, because we have no idea how the content responds to size changes. 

The Level Up content is actually a good case in point of that. For some reason, enlarging the canvas by 100-200 pixels causes it to zoom the content considerably. I'll see if there's a way to address that. Setting width instead of height could be an option, but I also want to make sure we don't end up shrinking the content too much.

I think ultimately what we want is to have the ability to specify a desired canvas size for the HTML5 app, and maybe have some auto-detection logic in ricecooker and Studio that can offer suggestions based on the content, but I believe that would require us to finish up the "extra_fields" support on ContentNodes?

indirectlylit commented:
Maybe we could define a 'size' setting on the content node, which can be set by a sushi chef or, longer-term, in the studio UI? For example, it could be set to one of:

* fixed height (in px)
* fixed aspect ratio (fraction or float)
* dynamically resize height

In all cases, it would expand to 100% of the available width

rtibbles commented:
To add my two cents - I think @kollivier and @indirectlylit are both right here. We should add a `size` attribute within `extra_fields`. As @kollivier notes, we have not implemented `extra_fields` on ContentNodes yet, so we would need to do that.

indirectlylit commented:
closing as superseded by https://github.com/learningequality/kolibri/pull/6873
